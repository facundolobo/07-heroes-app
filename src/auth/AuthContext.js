import { createContext } from "react";

export const AuthContext = createContext(); //creacion de context