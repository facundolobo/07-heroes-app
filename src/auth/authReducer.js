import { types } from "../types/types";

// const state={
//     name: 'Fernando',
//     logged:true
// }
export const authReducer = (state = {} ,action ) =>{ //cambiamos el estado segun la variable types

    switch (action.type) {

        case types.login:
            return { 
                ...action.payload,
                logged: true
            }
        case types.logout:
            return {
                logged: false
            }
    
        default:
            return state;
    }
}