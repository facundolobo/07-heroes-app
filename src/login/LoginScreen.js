import React from 'react'

export const LoginScreen = ({history}) => { //esta en las props del componente
    
    const handleLogin=()=>{
        //history.push('/'); //redirecciono a la pag x defecto  q es marvel
        history.replace('/'); //reemplazo en la historia que no visito el login
        //console.log('click');
    }

    
    return (
        <div className="conteiner m-5">
            <h1>LoginScreen</h1>
            <hr/>

            <button
                className="btn btn-primary"
                onClick={handleLogin}
            >
                Login
            </button>
        </div> 
        
    )
}
