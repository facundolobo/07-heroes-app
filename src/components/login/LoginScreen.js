import React, { useContext } from 'react'
import { AuthContext } from '../../auth/AuthContext';

import { types } from '../../types/types';

export const LoginScreen = ({history}) => { //esta en las props del componente
  
    //obtiene el ultimo lugar q ingreso, si no existe sera a por defecto 
    

    
    const {dispatch} = useContext(AuthContext)

    

    const handleLogin=()=>{

        const lastPath = localStorage.getItem('lastPath') || '/'; 
        //history.push('/'); //redirecciono a la pag x defecto  q es marvel
        //history.replace('/'); //reemplazo en la historia que no visito el login
        //console.log('click');
        // {
        //     name: 'Facundo'
        // }
        
        dispatch( {
            type:types.login,
            payload:{
                     name: 'Facundo'
                    }
        } )
        history.replace(lastPath); //enviamos ultima direccion al logearse
    }

    
    return (
        <div className="conteiner m-5">
            <h1>LoginScreen</h1>
            <hr/>

            <button
                className="btn btn-primary"
                onClick={handleLogin}
            >
                Login
            </button>
        </div> 
        
    )
}
