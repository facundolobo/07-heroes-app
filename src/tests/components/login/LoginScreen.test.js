import { mount } from "enzyme"
import { AuthContext } from "../../../auth/AuthContext"
import { LoginScreen } from "../../../components/login/LoginScreen"
import { types } from "../../../types/types"

describe('Pruebas en <LoginScreen/>', () => {
    //history
    const history={
        
        replace: jest.fn()
    }

    //contexto
    const contextValue ={
        dispatch: jest.fn(), //para indicar si fue llamada con que argumento etc
        user: {
            logged:false,
           
        }
    }

    const wrapper= mount(
        <AuthContext.Provider value={contextValue}>

            <LoginScreen history={ history }/>

        </AuthContext.Provider>
    )

    test('debe de mostrarce correctamente ', () => {
        //snapshot
        expect(wrapper).toMatchSnapshot();
    })


    test('debe realizar el dispatch y la navegaciÃ³n', () => {
        //simular click
        const handleLogin =wrapper.find('button').prop('onClick');

        handleLogin();


        //verificamos si  fue llamada 
        expect(contextValue.dispatch).toBeCalledWith({ //este objeto esta definido en "handleLogin"
            type:types.login,
            payload:{
                     name: 'Facundo' 
                    }
        })
        //verificar que repalce de history haya sido llamado con '/'
        expect(history.replace).toBeCalledWith('/');

        ////verificar que repalce de history haya sido llamado con '/dc' porque cambiamos localStore
        localStorage.setItem('lastPath','/dc')// cambiamos localStore
        
        handleLogin(); //simular click

        //verificar que repalce de history haya sido llamado con '/dc' porque cambiamos localStore
        expect(history.replace).toBeCalledWith('/dc');

    })
})
