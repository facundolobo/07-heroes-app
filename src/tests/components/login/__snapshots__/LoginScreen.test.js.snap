// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoginScreen/> debe de mostrarce correctamente  1`] = `
<div
  className="conteiner m-5"
>
  <h1>
    LoginScreen
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
